browser/ui/views/toolbar/wallet_button_notification_source.cc

```diff
@@ -0,0 +1,113 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/ui/views/toolbar/wallet_button_notification_source.h"
+
+#include "brave/browser/brave_wallet/keyring_service_factory.h"
+#include "brave/browser/brave_wallet/tx_service_factory.h"
+#include "brave/components/brave_wallet/browser/brave_wallet_utils.h"
+#include "brave/components/brave_wallet/browser/pref_names.h"
+
+namespace brave {
+
+WalletButtonNotificationSource::WalletButtonNotificationSource(
+    Profile* profile,
+    WalletButtonNotificationSourceCallback callback)
+    : callback_(callback) {
+  prefs_ = profile->GetPrefs();
+  tx_service_ = brave_wallet::TxServiceFactory::GetServiceForContext(profile);
+  keyring_service_ =
+      brave_wallet::KeyringServiceFactory::GetServiceForContext(profile);
+  tx_service_->AddObserver(tx_observer_.BindNewPipeAndPassRemote());
+  keyring_service_->AddObserver(
+      keyring_service_observer_.BindNewPipeAndPassRemote());
+  CheckTxStatus();
+}
+
+WalletButtonNotificationSource::~WalletButtonNotificationSource() {}
+
+void WalletButtonNotificationSource::MarkWalletButtonWasClicked() {
+  prefs_->SetBoolean(kWalletButtonClicked, true);
+  NotifyObservers();
+}
+
+void WalletButtonNotificationSource::CheckTxStatus() {
+  if (!tx_service_) {
+    NOTREACHED();
+    return;
+  }
+  status_resolver_ =
+      std::make_unique<brave_wallet::TxStatusResolver>(tx_service_);
+  status_resolver_->GetPendingTransactionsCount(
+      base::BindOnce(&WalletButtonNotificationSource::OnTxStatusResolved,
+                     weak_ptr_factory_.GetWeakPtr()));
```

Comment:
I would prefer we just implement and expose a new method `GetPendingTransactionsCount` in `TxService` interface to get the count for all chains in one mojo call. Optionally, we can also have an optional argument for specifying chains which callers are interested in that method, but we don't have the immediate usage yet so this can be done later.